modules/database/database.bal

```diff
@@ -899,21 +932,56 @@ public isolated transactional function getTimelineSubstepWithParams(int stepId,
     }
 }
 
-public isolated transactional function createTimelineSubstep(int stepId, string href, string? hrefUi, string? substepId) returns error? {
-    if href == "" {
+public isolated transactional function createTimelineSubstep(int stepId, TimelineSubstep substep) returns error? {
+    if substep.href == "" {
         return error("Href cannot be empty!");
     }
     int count = check experimentDB->queryRow(`SELECT count(*) FROM TimelineSubstep WHERE stepId=${stepId};`);
     count += 1;
+    string? substepId = substep.substepId;
     var insertResult = check experimentDB->execute(
         `INSERT INTO TimelineSubstep (stepId, substepNr, substepId, href, hrefUi, cleared) 
-         VALUES (${stepId}, ${count}, ${substepId != () ? substepId : count.toString()}, ${href}, ${hrefUi});`
+         VALUES (${stepId}, ${count}, ${substep.substepId != () ? substepId : count.toString()}, ${substep.href}, ${substep.hrefUi}, ${substep.cleared});`
     );
 }
 
-public isolated transactional function updateTimelineProgress(int stepId, float? progressStart, float? progressTarget, float? progressValue, string? progressUnit) returns error? {
+# Checks if substep is already in database and updates cleared field in case it is. Otherwise calls createTimelineSubstep. 
+#
+# + stepId - stepId
+# + substep - substep
+# + return - error or ()
+public isolated transactional function updateTimelineSubstep(int stepId, TimelineSubstep substep) returns error? {
+    TimelineSubstepSQL[] substeps = check getTimelineSubsteps(stepId);
```

Comment:
The low level update method should not have these kinds of high level safety checks with side-effects (mainly for performance in this case). The check should be done one level higher where the full list of substeps is available to be compared with the list of substeps from the DB. This should allow O(n) complexity for the check with only one extra DB query. This check can be implemented as an extra utility method. That method could also reside in the database module.